### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by dominictarr**
I had to refactor this to fully understand what was going on.

This change separates the part that manages the state of the widget from the part that suggests completions.

I made one small breaking change - the suggest box is triggered by a sigil, but then when something is selected it matches back towards the nearest sigil and then replaces that with the selected item, but now, the value should include the sigil, and suggest-box replaces the whole word.

It would be easy to add tests for the logic now, because box only touches a few aspects of the textarea (value, selectionStart), if the bits that manage it as an actual dom element (events and positioning) where moved, which would be easy.

This is needed for my naming pr https://github.com/ssbc/patchwork/pull/251

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -102,126 +182,27 @@ function deactivate() {
...
-
-
-function compare(a, b) {
-  return compareval(a.rank, b.rank) || compareval(a.title, b.title)
-}
-
-function compareval(a, b) {
-  return a === b ? 0 : a < b ? -1 : 1
+  var word = this.suggest(function (_, suggestions) {
+    console.log('SUG', _, suggestions)
```
Comment (pfrazee) [inline index.js]: log leftover

**Comment by pfrazee** [comment]
:+1: looks like a good improvement

**Comment by dominictarr** [comment]
fixed. if you pub with a major change I'll make a quick PR to get it into patchwork master (aside from names plugin is gonna take more disucssion)

**Comment by pfrazee** [comment]
oh did the readme get updated?

**Comment by dominictarr** [comment]
ah, no it didn't okay I'll fix that.

**Comment by pfrazee** [comment]
published `2.0.0`

**Pull request merged**